dialer_test.go

```diff
@@ -519,6 +517,196 @@ func TestConnectState(main *testing.T) {
+		assertUnreadyNotify(t, newUnreadyCh, "the error")
+
+		d.Close()
+
+		expected := `[DEBUG] connection unready
+[DEBUG] dialing
+[DEBUG] connection unready: the error
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("ReadyIfConnected", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		l := logger.NewTest()
+
+		readyCh := make(chan struct{}, 1)
+		unreadyCh := make(chan error, 1)
+
+		conn := mock_amqpextra.NewMockAMQPConnection(ctrl)
+		conn.EXPECT().NotifyClose(any()).AnyTimes()
+		conn.EXPECT().Close().AnyTimes()
+
+		d, err := amqpextra.NewDialer(
+			amqpextra.WithURL("amqp://rabbitmq.host"),
+			amqpextra.WithLogger(l),
+			amqpextra.WithAMQPDial(
+				amqpDialStub(conn),
+			),
+		)
+		require.NoError(t, err)
+		newReadyCh, newUnreadyCh := d.Notify(readyCh, unreadyCh)
+
+		assertUnreadyNotify(t, newUnreadyCh, amqp.ErrClosed.Error())
+		assertReadyNotify(t, newReadyCh)
+
+		d.Close()
+
+		expected := `[DEBUG] connection unready
+[DEBUG] dialing
+[DEBUG] connection ready
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("UnreadyWhileWaitRetry", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		l := logger.NewTest()
+
+		readyCh := make(chan struct{}, 1)
+		unreadyCh := make(chan error, 1)
+
+		conn := mock_amqpextra.NewMockAMQPConnection(ctrl)
+		conn.EXPECT().NotifyClose(any()).AnyTimes()
+		conn.EXPECT().Close().AnyTimes()
+
+		d, err := amqpextra.NewDialer(
+			amqpextra.WithURL("amqp://rabbitmq.host"),
+			amqpextra.WithLogger(l),
+			amqpextra.WithAMQPDial(
+				amqpDialStub(errors.New("the error")),
+			),
+			amqpextra.WithRetryPeriod(time.Millisecond*50),
+		)
+		require.NoError(t, err)
+		_, newUnreadyCh := d.Notify(readyCh, unreadyCh)
+
+		assertUnreadyNotify(t, newUnreadyCh, amqp.ErrClosed.Error())
+		d.Close()
+		time.Sleep(time.Millisecond * 20)
+
+		expected := `[DEBUG] connection unready
+[DEBUG] dialing
+[DEBUG] connection unready: the error
+[DEBUG] connection closed
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("UnreadyAfterClosed", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		l := logger.NewTest()
+
+		readyCh := make(chan struct{}, 1)
+		unreadyCh := make(chan error, 1)
+
+		conn := mock_amqpextra.NewMockAMQPConnection(ctrl)
+		conn.EXPECT().NotifyClose(any()).AnyTimes()
+		conn.EXPECT().Close().AnyTimes()
+
+		d, err := amqpextra.NewDialer(
+			amqpextra.WithURL("amqp://rabbitmq.host"),
+			amqpextra.WithLogger(l),
+			amqpextra.WithAMQPDial(
+				amqpDialStub(time.Millisecond*10, conn),
+			),
+		)
+		require.NoError(t, err)
+		newReadyCh, newUnreadyCh := d.Notify(readyCh, unreadyCh)
+
+		assertUnreadyNotify(t, newUnreadyCh, amqp.ErrClosed.Error())
+		time.Sleep(time.Millisecond * 20)
+		assertReadyNotify(t, newReadyCh)
+		d.Close()
```

Comment:
the purpose of the test is to cover the case when d.Notify() is called after d.Close()